File: 백준/Gold/17265. 나의 인생에는 수학과 함께/나의 인생에는 수학과 함께.java
import java.util.*;
import java.io.*;

class Main {
	
	// IO field
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
	static StringTokenizer st;

	static void nextLine() throws Exception {st = new StringTokenizer(br.readLine());}
	static int nextInt() {return Integer.parseInt(st.nextToken());}
	static long nextLong() {return Long.parseLong(st.nextToken());}
	static void bwEnd() throws Exception {bw.flush();bw.close();}
	
	// Additional field

	static int N;
	static char[][] A;
	static int low = Integer.MAX_VALUE;
	static int high = Integer.MIN_VALUE;
	
	public static void main(String[] args) throws Exception {
		
		ready();
		solve();
	
		bwEnd();
		
	}
	
	static void ready() throws Exception{
		
		N = Integer.parseInt(br.readLine());
		A = new char[N][N];
		for(int i=0;i<N;i++) {
			String temp = br.readLine();
			for(int j=0;j<2*N;j+=2)A[i][j>>1] = temp.charAt(j);
		}
		
	}
	
	static void solve() throws Exception{
		
		bck(0,0,A[0][0]-'0');
		bw.write(high + " " + low);
		
	}
	
	static void bck(int x, int y, int res) {
		
		if(x == N-1 && y == N-1) {
			low = Math.min(low, res);
			high = Math.max(high, res);
			return;
		}
		
		if(x < N-1) {
			if(x < N-2) bck(x+2,y,cal(res,A[x+1][y],A[x+2][y]-'0'));
			if(y < N-1) bck(x+1,y+1,cal(res,A[x+1][y],A[x+1][y+1]-'0'));
		}
		
		if(y < N-1) {
			if(y < N-2) bck(x,y+2,cal(res,A[x][y+1],A[x][y+2]-'0'));
			if(x < N-1) bck(x+1,y+1,cal(res,A[x][y+1],A[x+1][y+1]-'0'));
		}
	}
	
	static int cal(int a, char op, int b) {
		if(op == '+') return a+b;
		if(op == '-') return a-b;
		return a*b;
	}
	
}

File: 백준/Gold/17265. 나의 인생에는 수학과 함께/나의 인생에는 수학과 함께.cpp
#include <iostream>
using namespace std;

char map[5][5]{};
int mx = -15626, mn = 15626, n;

void dfs(int x, int y, int v, char o) {
	if (o == 'o') {
		if (x + 1 < n)	dfs(x + 1, y, v, map[x][y]);
		if (y + 1 < n)	dfs(x, y + 1, v, map[x][y]);
	}
	else {
		int t = map[x][y] - '0';
		if (x || y) {
			if (o == '+')	v += t;
			if (o == '-')	v -= t;
			if (o == '*')	v *= t;
			if (x == n - 1 && y == n - 1) {
				mx = max(mx, v);
				mn = min(mn, v);
				return;
			}
		}
		else	v = t;
		if (x + 1 < n)	dfs(x + 1, y, v, 'o');
		if (y + 1 < n)	dfs(x, y + 1, v, 'o');
	}
}

int main() {
	cin >> n;
	for (int i = 0; i < n; i++)
		for (int j = 0; j < n; j++)
			cin >> map[i][j];
	dfs(0, 0, 0, '+');
	cout << mx << ' ' << mn;
}